### ISSUE TITLE
Set culture from arguments?

### ISSUE THREAD

**Issue by PhilippeWaeber** [EntityGraphQL/EntityGraphQL#372]
We've got a generic localized record, where `Current` returns the localized value (using a fallback, if the wanted value is `null`) for a the current language:

```c#
public abstract record Localized<T>
{
    public required T De { get; set; }
    public T? Fr { get; set; }
    public T? It { get; set; }
    public T? En { get; set; }

    public T Current =>
        CultureInfo.CurrentCulture.TwoLetterISOLanguageName switch
        {
            "de" => De,
            "fr" => Fr ?? De,
            "it" => It ?? Fr ?? De,
            "en" => En ?? De,
            _ => De
        };
}
```

I've intended to use this with request localization, using the `Accept-Language` HTTP header, registered like this during startup:
```c#
app.UseRequestLocalization(options =>
{
    var supportedCultureInfos = supportedCultures.Select(culture => new CultureInfo(culture))
                                                 .ToList();

    options.SupportedCultures = supportedCultureInfos;
    options.SupportedUICultures = supportedCultureInfos;

    options.DefaultRequestCulture = new RequestCulture(supportedCultureInfos.First());

    // Only determine language according to the Accept-Language header
    options.RequestCultureProviders.Clear();
    options.RequestCultureProviders.Add(new AcceptLanguageHeaderRequestCultureProvider());
});
```

Our frontend guys wish to be able to specify the language using an argument. E.g.
```graphql
products (culture: "fr") {
  description { current }
}
```
Where description.current should contain the french text (or fallback if it is null).
Ideally, this argument could be specified on all fields.

To achieve this, without having to modify the existing implementation, I'd like to set the culture like this:
```cs
Thread.CurrentThread.CurrentCulture = new CultureInfo(args.culture!);
Thread.CurrentThread.CurrentUICulture = new CultureInfo(args.culture!);
```
But where to do that best? 
- Use request localization with a custom culture provider?
- Did I miss a possibility to specify such a behavior during the field creation?

**Comment by lukemurray** [comment]
Sorry I am rarely had to work with localization!

Thinking off the top of my head (i.e. none of these may be a good way). Does that `Localized<T>` work with EF? Or is this against data in memory?

You (maybe) could defined the description fields as below and modify the `Localized<T>` to take the argument.

```cs
schemaType.AddField("description",
    new {
        culture = "fr" // or some default
    },
    (t, args) => t.Description.Current(args.culture),
    "Get description in defined culture"
);
```

One way to get access to that argument in a way you can then set the thread is with a service field. But without knowing more about your set up I don't know how this will work or scale.

```cs
schemaType.AddField("description",
    new {
        culture = "fr" // or some default
    },
    "Get description in defined culture"
).Resolve<ILocalizedService>((t, args, service) => service.GetDescription(t.Description, args.culture));
```

You'd then have `LocalizedService` set the thread culture and call `.Current` and maybe set the thread culture back.

You could also use a custom directive. Then the GQL request would be like

```gql
products {
  description @culture(lang: "fr")
}
```

And you'd be able to set the thread culture in the directive.

I hope this helps you find a way.

**Comment by PhilippeWaeber** [comment]
Maybe I missed something but I don't think this fits what I'd like to do.
My issue is not really related to localization but rather that I'd like to perform an action **after** the graphql request has been deserialized but **before** the expression is executed (against EF in my case).

To illustrate what I'd imagine, I've added `Action<TParams> setupAction` parameter to `ReplaceField`:
```cs
 public Field ReplaceField<TParams, TReturn>(string name, TParams argTypes, Expression<Func<TBaseType, TParams, TReturn>> fieldSelection, Action<TParams> setupAction, string? description)
 {
   [...]
}
```
The `setupAction` is not used or forwarded for now, I wanted to check if a construct like this would make sense to you as well, before trying to figure this out.

to be used like this during field setup:
```cs
schema.ReplaceField("substances",
                    new
                    {
                        SearchTerm = (string?)null,
                        Culture = "de"
                    },
                    (ctx, args) => ctx.Substances
                                      .WhereWhen(SearchExpressions.SearchSubstances(args.SearchTerm!),
                                                 !string.IsNullOrWhiteSpace(args.SearchTerm) && args.SearchTerm.Length >= 3),
                    args =>
                    {
                        Thread.CurrentThread.CurrentCulture = new CultureInfo(args.Culture);
                        Thread.CurrentThread.CurrentUICulture = new CultureInfo(args.Culture);
                    },
                    "Returns a list of substances.")
      .UseSort((Substance substance) => new
               {
                   substance.Id,
                   substance.SubstanceNumber,
                   DescriptionDe = substance.Description.De
               },
               substance => substance.SubstanceNumber,
               SortDirection.ASC
      )
      .UseOffsetPaging(GraphQLConstants.Paging.DefaultPageSize, GraphQLConstants.Paging.MaxPageSize);
```

**Comment by lukemurray** [comment]
More likely I am not following exactly what you're asking :) 

I'm not sure in your example how you want to use `configureService` as you don't use it in the second code sample. If you need to have access to a service within a field you can use the [`Resolve` methods](https://entitygraphql.github.io/docs/schema-creation/other-data-sources). Note they get executed after EF by default.

Regarding your comment about executing something after the request has been processed but before the execution, currently the only option is `ExecutionOptions.BeforeExecuting`. Here is the doc code

```
        /// <summary>
        /// Allows you to hook into just before an expression is executed and modify it to suit. Note that if 
        /// <code>ExecuteServiceFieldsSeparately</code> is true, this will be called twice if your query includes fields with services.
        /// Second parameter bool isFinal == true if the expression is the final execution - this means
        ///  - ExecuteServiceFieldsSeparately = false, or
        ///  - The query does not reference any fields with services
        ///  - The query references fields with service and the first execution has completed (isFinal == false) and we are executing again to merge the service results
        /// </summary>
```

You'll see you do not have access to the arguments.

**Comment by PhilippeWaeber** [comment]
I do not need to modify the expression :)

Maybe an example will illustrate what I'm trying to achieve.

I've got a `Product`, having a description, being a `LocalizedText`

```cs
public class Product : BaseEntity
{
    public int ProductNumber { get; set; }

    public LocalizedText Description { get; set; } = LocalizedText.Default;
}
```

Where `LocalizedText` is
```cs
public record LocalizedText : Localized<string>
{
    public static LocalizedText Default => new() { De = "{default}" };
}
```

And `Localized<T>` is
```cs
public abstract record Localized<T>
{
    public required T De { get; set; }
    public T? Fr { get; set; }
    public T? It { get; set; }
    public T? En { get; set; }

    public T Current =>
        CultureInfo.CurrentCulture.TwoLetterISOLanguageName switch
        {
            "de" => De,
            "fr" => Fr ?? De,
            "it" => It ?? Fr ?? De,
            "en" => En ?? De,
            _ => De
        };
}
```

When querying this model with this GraphQL query
```graphql
{
  products {
    totalItems
    items {
      id
      productNumber
      description {
        current
      }
    }
  }
}
```

The expression is translated and following SQL is executed:
```sql
SELECT TOP(@__p_0) [p].[Id], [p].[ProductNumber], CAST(0 AS bit), [p].[Description_De], [p].[Description_En], [p].[Description_Fr], [p].[Description_It], [p].[PeriodEnd], [p].[PeriodStart]
FROM [Products] AS [p]
ORDER BY [p].[ProductNumber]
```
Note that I request `Current`, which results in EF automagically querying the four languages it might be using. 
The actual resolution of the value of `Current` is done in memory. 

This is why I'd like to be able to set the culture of the thread, according to the graphql argument.

> I'm not sure in your example how you want to use configureService as you don't use it in the second code sample.

I don't have a use case for this just yet, we can forget about this for now. Just thought it might come in handy to be able to access services for some scenarios.

**Comment by lukemurray** [comment]
Yeh ok. thanks for that. Make sense to me know.

Currently you do not have access to the arguments for that field on the `BeforeExecuting` callback. And if we made that change I'd need to think about what that means as args are per field, not per executing thread. you could have multiple args requesting different cultures.

**Comment by PhilippeWaeber** [comment]
Yep, I had the same thoughts when digging into this and doing some debugging. At first, I imagined having a scope like the request scope of asp.net, for a single request. Until I've figured we might have multiple fields in a single query (for now that's how we use our API), which would require having to reset the state after each field handling, which can't be done automatically and would require a setup/teardown. 

With hindsight this doesn't sound like a good idea.
